### PR TITLE
Use canister outcalls for http execution

### PR DESCRIPTION
**Pull request by zvolin**

### CODE REVIEW

File: execution/Cargo.toml [hunk 1]
```diff
@@ -21,6 +21,8 @@ thiserror = "1.0.37"
 
 common = { path = "../common" }
 consensus = { path = "../consensus" }
+ic-http = { path = "../ic-http"}
+ic-cdk = "0.10.0"
```
Comment (oblique) [inline execution/Cargo.toml]: IMO this should be target guarded, unless if this is changed: https://github.com/eigerco/helios/blob/e42c2cda990f81c9e82aa300d96126b4f124f972/ic-http/src/http_request.rs#L14C1-L19

File: execution/src/rpc/http_rpc.rs [hunk 2]
```diff
@@ -79,91 +70,60 @@ impl ExecutionRpc for HttpRpc {
...
-        let tx = self
-            .provider
-            .send_raw_transaction(bytes)
-            .await
-            .map_err(|e| RpcError::new("send_raw_transaction", e))?;
 
-        Ok(tx.tx_hash())
-        */
-        todo!()
+        Ok(self
```
Comment (oblique) [inline execution/src/rpc/http_rpc.rs:97]: I think you need to call `serde_json::to_value` before this

File: execution/src/rpc/http_rpc.rs [hunk 3]
```diff
@@ -172,14 +132,101 @@ impl ExecutionRpc for HttpRpc {
...
+            .url(self.url.as_ref())
+            .method(HttpMethod::POST)
+            .header(HttpHeader {
+                name: "Content-Type".to_owned(),
+                value: "application/json".to_owned(),
+            })
+            .body(serde_json::to_vec(&payload)?)
+            .build();
+
+        let response = ic_http::http_request(request, 100_000_000)
```
Comment (oblique) [inline execution/src/rpc/http_rpc.rs]: Add a `TODO` here about the cycles

File: execution/Cargo.toml (hunk 1 above)
Comment (zvolin) [reply execution/Cargo.toml]: removed ic-http, left ic-cdk unguarded

File: execution/src/rpc/http_rpc.rs (hunk 2 above)
Comment (zvolin) [reply execution/src/rpc/http_rpc.rs:97]: manual calls to `to_value` are needed only if there is more than one value, to have uniform types in array

File: execution/Cargo.toml [hunk 4]
```diff
@@ -21,7 +21,6 @@ thiserror = "1.0.37"
 
 common = { path = "../common" }
 consensus = { path = "../consensus" }
-ic-http = { path = "../ic-http"}
 ic-cdk = "0.10.0"
```
Comment (oblique) [inline execution/Cargo.toml]: target guard `ic-cdk`

File: common/Cargo.toml [hunk 5]
```diff
@@ -11,5 +11,8 @@ hex = "0.4.3"
 ssz-rs = { git = "https://github.com/ralexstokes/ssz-rs", rev = "d09f55b4f8554491e3431e01af1c32347a8781cd" }
 ethers-core = "2.0.8"
 thiserror = "1.0.37"
+candid = "0.9.1"
```
Comment (oblique) [inline common/Cargo.toml]: remove candid and use [this](https://docs.rs/ic-cdk/latest/ic_cdk/api/management_canister/http_request/fn.http_request.html)

File: common/Cargo.toml [hunk 6]
```diff
@@ -11,5 +11,8 @@ hex = "0.4.3"
...
 ic-cdk = "0.10.0"
```
Comment (oblique) [inline common/Cargo.toml:16]: target guard `ic-cdk`

File: common/src/http.rs [hunk 7]
```diff
@@ -1,9 +1,156 @@
...
+// Values are taken from https://internetcomputer.org/docs/current/developer-docs/gas-cost
+// for a 13-node app subnet.
+pub const HTTP_OCALL_BASE_COST: u128 = 49_140_000;
+pub const HTTP_OCALL_REQ_BYTE_COST: u128 = 5_200;
+pub const HTTP_OCALL_RESP_BYTE_COST: u128 = 10_400;
+
+// TODO: should we pass cycles in http_x methods or we should have a default one?
+pub const DEFAULT_HTTP_OUTCALL_COST: u128 = HTTP_OCALL_BASE_COST
+    + 5 * 1024 * HTTP_OCALL_REQ_BYTE_COST
+    + 5 * 1024 * HTTP_OCALL_RESP_BYTE_COST;
```
Comment (oblique) [inline common/src/http.rs]: move this under `mod icp`

File: common/src/http.rs [hunk 8]
```diff
@@ -1,9 +1,156 @@
...
+            method: HttpMethod::POST,
+            headers,
+            body: Some(body),
+            ..Default::default()
+        };
+        let resp = ic_http_outcall(req, DEFAULT_HTTP_OUTCALL_COST).await?;
+        resp.0.try_into()
+    }
+
+    async fn ic_http_outcall(
```
Comment (oblique) [inline common/src/http.rs]: remove this and replace it with [`http_request`](https://docs.rs/ic-cdk/latest/ic_cdk/api/management_canister/http_request/fn.http_request.html)

File: execution/Cargo.toml (hunk 4 above)
Comment (zvolin) [reply execution/Cargo.toml]: done

File: common/Cargo.toml (hunk 5 above)
Comment (zvolin) [reply common/Cargo.toml]: done

File: execution/Cargo.toml (hunk 4 above)
Comment (zvolin) [reply execution/Cargo.toml]: oops, not done

File: common/src/http.rs (hunk 7 above)
Comment (zvolin) [reply common/src/http.rs]: done

File: common/src/http.rs (hunk 8 above)
Comment (zvolin) [reply common/src/http.rs]: done

File: common/Cargo.toml (hunk 6 above)
Comment (zvolin) [reply common/Cargo.toml:16]: done

File: execution/Cargo.toml (hunk 4 above)
Comment (zvolin) [reply execution/Cargo.toml]: removed completely

**Review by oblique — APPROVED** [review]
LGTM

**Pull request merged**